server/src/internalClusterTest/java/org/opensearch/clustermanager/ClusterManagerTaskThrottlingIT.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.clustermanager;
+
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.admin.cluster.settings.ClusterUpdateSettingsRequest;
+import org.opensearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.opensearch.cluster.metadata.ProcessClusterEventTimeoutException;
+import org.opensearch.cluster.service.ClusterManagerThrottlingException;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.test.OpenSearchIntegTestCase;
+import org.opensearch.transport.TransportService;
+import org.opensearch.transport.TransportMessageListener;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.SUITE, numDataNodes = 0)
+public class ClusterManagerTaskThrottlingIT extends OpenSearchIntegTestCase {
+
+    /*
+     * This integ test will test end-end cluster manager throttling feature for
+     * remote master.
+     *
+     * It will check the number of request coming to master node
+     * should be total number of requests + throttled requests from master.
+     * This will ensure the end-end feature is working as master is throwing
+     * Throttling exception and data node is performing retries on it.
+     *
+     */
+    public void testThrottlingForRemoteMaster() throws Exception {
+        try {
+            internalCluster().beforeTest(random());
+            String clusterManagerNode = internalCluster().startClusterManagerOnlyNode();
+            String dataNode = internalCluster().startDataOnlyNode();
+            int throttlingLimit = randomIntBetween(1, 5);
+            createIndex("test");
+            setPutMappingThrottlingLimit(throttlingLimit);
+
+            TransportService masterTransportService = (internalCluster().getInstance(TransportService.class, clusterManagerNode));
+            AtomicInteger requestCountOnMaster = new AtomicInteger();
+            AtomicInteger throttledRequest = new AtomicInteger();
+            int totalRequest = randomIntBetween(throttlingLimit, 5 * throttlingLimit);
+            CountDownLatch latch = new CountDownLatch(totalRequest);
+
+            masterTransportService.addMessageListener(new TransportMessageListener() {
+                @Override
+                public void onRequestReceived(long requestId, String action) {
+                    if (action.contains("mapping")) {
+                        requestCountOnMaster.incrementAndGet();
+                    }
+                }
+
+                @Override
+                public void onResponseSent(long requestId, String action, Exception error) {
```

Comment:
how are we making sure that the requests are indeed throttled here ? Don't we need to create our own Action which is blocking in nature to induce throttling ?

Reply:
Here my intention was to test the end-end functionality how cx uses it. 
Here we are setting threshold limits for put-mapping tasks and generates the more requests then the limit. We are verifying if there are any exceptions coming from master node it is throttling exceptions only, and those exceptions are not getting exposed to cx and being retried by the data node and final successful response is returned to the cx.


> How are we making sure that the requests are indeed throttled here ?

We are relying on random scenarios to test every possible case. We are not ensuring throttling every time, sometime there wont be throttling as well. Idea is test case should pass every time.

> Don't we need to create our own Action which is blocking in nature to induce throttling ?

We have done that and verified in the Unit tests.

Reply:
Can we run the ITs multiple times and see there is some throttling  ? if there is no throttling, this test wouldn't serve its purpose. 

> We have done that and verified in the Unit tests.

Doing similar thing in ITs would be an add on as we are verifying more things here like total retries etc.

Reply:
There is one IT where we are doing throttling for all the calls.(testTimeout one) That would serve that usecase.